runtime/src/qio/qio_formatted.c

```diff
@@ -2198,24 +2200,122 @@ int _ltoa(char* restrict dst, size_t size, uint64_t num, int isnegative,
   return i;
 }
 
-// error codes:
+// Converts num to a string in buf, returns the number
+// of bytes that would be used if space permits (not including null)
+// or -1 on error
+// in *skip, returns the number of bytes at the front of buf that
+// should be skipped in the final output. Such bytes are counted
+// in the return value. (this happens with 0x in %a conversions)
+static
+int _ftoa_core(char* buf, size_t buf_sz, double num,
+               int base, int realfmt,
+               int precision, int uppercase,
+               int prefix_base,
+               int *skip)
+{
+  int got = 0;
+
+  // Do the numeric conversion and figure out how big the output
+  // is. This conversion must concern itself with precision but
+  // need not handle padding or various trailing . or .0 rules.
+  // That is handled in the caller.
+
+  *skip = 0;
+
+  if( base == 16 ) {
```

Comment:
Ah, I see that most of this code is just moved from another location.  I don't think it's necessary to add functionality that wasn't already present (if it isn't), but maybe we should add a note to indicate where it would go or something?